posts/2019/07/uarray-update-api-changes-overhead-and-comparison-to-__array_function__.ipynb

```diff
@@ -520,25 +548,316 @@
+    "                    return NotImplemented\n",
+    "\n",
+    "                converted.append({k: int(v) for k, v in d.value.items()})\n",
+    "            elif d.type is np.dtype:\n",
+    "                try:\n",
+    "                    converted.append(np.dtype(d.value))\n",
+    "                except (TypeError, ValueError):\n",
+    "                    return NotImplemented\n",
+    "            else:\n",
+    "                # Handle the base case, pass through everything else\n",
+    "                converted.append(d.value)\n",
+    "        return converted"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 23,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "<xarray.DataArray (t: 3, x: 5)>\n",
+      "array([[1., 1., 1., 1., 1.],\n",
+      "       [1., 1., 1., 1., 1.],\n",
+      "       [1., 1., 1., 1., 1.]])\n",
+      "Dimensions without coordinates: t, x\n"
+     ]
+    }
+   ],
+   "source": [
+    "with ua.set_backend(XarrayBackend):\n",
+    "    print(ones({'t': 3, 'x': 5}))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Note here how a shape as a dictionary was automatically intercepted by XArray, and creates a suitable `DataArray`. This would simply not be possible in `__array_function__`. Let's explore a more exotic case:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 24,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "<xarray.DataArray 'ones-c74937aa1cf5a1e9c9dd0ca4c337fd2c' (t: 3, x: 5)>\n",
+      "dask.array<shape=(3, 5), dtype=float64, chunksize=(3, 5)>\n",
+      "Dimensions without coordinates: t, x\n"
+     ]
+    }
+   ],
+   "source": [
+    "with ua.set_backend(DaskBackend), ua.set_backend(XarrayBackend):\n",
+    "    print(ones({'t': 3, 'x': 5}))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "And just like that, you can compose the two, without either having to be aware of the other!"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## The overhead of `uarray`\n",
+    "\n",
+    "One of the limiting factors in the adoption of `uarray` has been its overhead. This is currently [being brought down](https://github.com/Quansight-Labs/uarray/pull/170) by [Peter Bell](https://github.com/peterbell10), a GSoC student working on `uarray`, who is working on a C++ implementation of the protocol. As of the last version, the overhead on my machine was about 449 ns per call (or about 12.4 Python function calls), and this is still being worked on. You can follow the [updates on Peter’s work here](https://blogs.python-gsoc.org/en/blogs/peterbell10s-blog/). It is also being [merged into `scipy.fft`](https://github.com/scipy/scipy/pull/10383) as part of his ongoing GSoC work."
```

Comment:
That 12.4 Python function calls statement should probably be qualified as "do-nothing no-argument":
```
In [1]: def f(): pass                                                                          

In [2]: %timeit f()                                                                            
59.7 ns ± 0.0635 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [3]: def ones(shape, dtype=None, order='C'): pass                                           

In [4]: %timeit ones(None)                                                                     
81.9 ns ± 0.585 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```
Or instead say, "about the same level as the NumPy ufunc or `__array_function__` overhead"